source/slang/slang-check-expr.cpp

```diff
@@ -1355,6 +1355,162 @@ namespace Slang
         }
     }
 
+    RefPtr<Expr> SemanticsVisitor::CheckMatrixSwizzleExpr(
+        MemberExpr* memberRefExpr,
+        RefPtr<Type>      baseElementType,
+        IntegerLiteralValue baseElementRowCount,
+        IntegerLiteralValue baseElementColCount)
+    {
+        RefPtr<MatrixSwizzleExpr> swizExpr = new MatrixSwizzleExpr();
+        swizExpr->loc = memberRefExpr->loc;
+        swizExpr->base = memberRefExpr->baseExpression;
+
+        int elementIndices[4][2];
+        int elementCount = 0;
+
+        bool anyDuplicates = false;
+        int oneIndexed = -1;
+
+        String swizzleText = getText(memberRefExpr->name);
+        auto coursor = swizzleText.begin();
+
+        while (coursor != swizzleText.end())
```

Comment:
As it turns out here things work out ok - but it probably worth pointing out that this surrounding loop check is looking for end, but that checking is not honored by the subsequent code. Which can step 1 or more characters and therefore could in principal step past end. If you know the cursor is a character pointer you might be better off with < (although that does not fix the problem). This works out because as it turns out the buffer held by String is 0 terminated, and so the subsequent code will hit that, and report as an 'invalid character'. 
That's probably not quite the error you want. 
Note that UnownedStringBuffer is not zero terminated, and so bad things would happen. 

So how might you improve this? Doing this kind of char checking whilst constantly checking for end can be a little tedious. Here you know it's zero terminated, that being the case you probably don't want to mix the mechanisms. 

More concretely something like

```
// the contents of String are zero terminated
const char* cursor = swizzleText.getBuffer(); 
while (*cursor) 
{ 
    ... 

```

Reply:
FWIW, I specifically advocated to @Checkmate50 to use this style and rely on the fact that the buffer for a `String` is null-terminated. I believe the error messages won't specifically call out a single invalid character, but will instead reference the entire swizzle string as being invalid. I might be wrong though, and you may have looked at the code more closely than I have.

Reply:
The issue wasn't so much about relying on zero termination - that makes a lot of sense for this style of lexing. I realize my comment might waffle around this - as I'm trying to figure out what's going on. The problem is the *mix* where the outside loop is written to use the start/end style, which implies that the remaining code is follows the same mechanism which it doesn't, so that raises alarms. So just making the loop rely an the the same zero termination style would make it ok.